src/util/logging.cpp

```diff
@@ -17,6 +16,18 @@
 #include "controllers/controllerdebug.h"
 #include "util/assert.h"
 
+Q_LOGGING_CATEGORY(mixxxLog, MIXXX_LOGGING_CATEGORY_ROOT)
+
+Q_LOGGING_CATEGORY(mixxxLogConfig, MIXXX_LOGGING_CATEGORY_CONFIG)
```

Comment:
Is there any advantage from defining them all in one place?
Personally, I think all these macros look ugly.

I'd prefer this:

> ```cpp
> bool Highscore::load(const QString &filepath)
> {
>     QLoggingCategory fcIo("fc.io");
>     qCDebug(fcIo) << "Loading highscore ...";
>     QFile file(filepath);
>     if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
>         qCWarning(fcIo) << "Could not open file " << filepath << "!";
>         return;
>     }
>     // ...
> }
>```
>
> [...]
>
> QLoggingCategory represents a category in code, in this case it's "fc.io". It's not a one to one mapping though ... you can have multiple instances representing the same category.

Source: https://www.qt.io/blog/2014/03/11/qt-weekly-1-categorized-logging

Reply:
The idea was to group the logs under some predefined top-level categories to allow basic filtering without needing to know all child categories. The custom macro MIXXX_LOGGING_CATEGORY_PARENT_CHILD establishes a naming convention.

The macros Q_DECLARE_LOGGING_CATEGORY/Q_LOGGING_CATEGORY help to find all categories that are available, even if they are only defined locally inside a .cpp file and not intended to be reused.

Reply:
> The idea was to group the logs under some predefined top-level categories to allow basic filtering without needing to know all child categories. The custom macro MIXXX_LOGGING_CATEGORY_PARENT_CHILD establishes a naming convention.

I thought this is done implicitly by separating the category name with dots.
And filtering can be done like this:

    QT_LOGGING_RULES="mixxx.engine.controls.*=true"

> The macros Q_DECLARE_LOGGING_CATEGORY/Q_LOGGING_CATEGORY help to find all categories that are available, even if they are only defined locally inside a .cpp file and not intended to be reused.

So you can list all categories via command line? Or do you mean by grepping through the source?

Reply:
@uklotzde can you elaborate?

Reply:
Ping.

Reply:
Without the constants all the logging categories would be scattered in the code as arbitrary, redundant strings. Renaming or restructuring them would be really tedious.

The Q_DECLARE_LOGGING_CATEGORY macro explicitly applies the *separated by dots* convention. Otherwise there is no rule how to structure the names of logging categories.

Reply:
OK, sounds good. I'll give this a test. Hopefully we can stop sprinkling commented out qdebug calls all over our code :smile:

Reply:
I don't like to introduce a custom macro. But so far I didn't have any idea how to achieve this in a better way.

Reply:
We could put all category strings in a single header file and then include them from there. That way you only need to seach-replace in a single file.

Reply:
With this restriction the macros could be avoided.

Reply:
After 2nd thought: Not viable. A huge global file that might contain a separate logging category for each module (comprising 1 or a few .cpp files) is unmaintainable.

How is a user supposed to chose the correct one? Even the list of predefined top-level logging categories in logging.h is long already.

Reply:
Hmm. I also thought about this and I'm not even sure we need this global list. When we define the logging category inside the source file, like this it becomes much easier to find the correct logging category to enable debug messages for:

    QLoggingCategory logcat("device.soundsource.portaudio");

You're right that it becomes a bit tedious to restructure logging categories later on if the fully qualified names are spread across the code base like this, but it's still possible to mass-replace them in a bash one-liner using `find`/`grep`/`sed`. And I don' t think we want to restructure them that often, and the instant visibility of the full logging category name when reading the source file is very convenient and outweighs this potential issue IMHO.